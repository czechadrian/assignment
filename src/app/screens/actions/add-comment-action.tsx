import { ActionType, createAction } from "typesafe-actions";
import { TComment } from "../../reducers/comments";

export const addCommentAction = createAction("comment/ADD_NEW_COMMENT")<
  TComment
>();

export type TAddCommentAction = ActionType<typeof addCommentAction>;
