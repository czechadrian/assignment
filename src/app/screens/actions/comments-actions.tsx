import { ActionType, createAction } from "typesafe-actions";
import { ThunkAction } from "redux-thunk";
import { TRootState } from "../../reducers";
import { getComments } from "../../../api-wrapper/get-comments";
import { TComments } from "../../reducers/comments";

export const getCommentsInitAction = createAction("data/FETCH_INIT")();
export const getCommentsSuccessAction = createAction("data/FETCH_SUCCESS")<
  Pick<TComments, "comments">
>();
export const getCommentsFailureAction = createAction("data/FETCH_FAILURE")();

export type TGetCommentsInitAction = ActionType<typeof getCommentsInitAction>;
export type TGetCommentsSuccessAction = ActionType<
  typeof getCommentsSuccessAction
>;
export type TGetCommentsFailureAction = ActionType<
  typeof getCommentsFailureAction
>;

export type TGetCommentsActions =
  | TGetCommentsSuccessAction
  | TGetCommentsFailureAction
  | TGetCommentsInitAction;

export type TFetchCommentsAction = () => ThunkAction<
  void,
  TRootState,
  null,
  TGetCommentsActions
>;
export const fetchComments: TFetchCommentsAction = () => (dispatch) => {
  dispatch(getCommentsInitAction());
  return getComments()
    .then((payload) => {
      return dispatch(getCommentsSuccessAction({ comments: payload }));
    })
    .catch(() => {
      dispatch(getCommentsFailureAction());
    });
};
