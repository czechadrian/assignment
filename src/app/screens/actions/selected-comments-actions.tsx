import { ActionType, createAction } from "typesafe-actions";
import { TComment } from "../../reducers/comments";

export const selectCommentAction = createAction(
  "homepage/CLICK_SELECT_COMMENT"
)<Pick<TComment, "id">>();
export const removeSelectedCommentAction = createAction(
  "selected-comments/REMOVE_SELECTED_COMMENT"
)<Pick<TComment, "id">>();

export type TSelectCommentAction = ActionType<typeof selectCommentAction>;

export type TRemoveSelectedCommentAction = ActionType<
  typeof removeSelectedCommentAction
>;

export type TSelectedCommentActions =
  | TSelectCommentAction
  | TRemoveSelectedCommentAction;
