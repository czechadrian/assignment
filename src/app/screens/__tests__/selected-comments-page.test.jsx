import React from "react";
import "jest-styled-components";
import renderer from "react-test-renderer";
import { useHistory } from "react-router-dom";
import SelectedCommentsPage from "../selected-comments-page";
import { useSelector, useDispatch } from "react-redux";
import { removeSelectedCommentAction } from "../actions/selected-comments-actions";
import { mount } from "enzyme";
const { payload } = require("../../../api-wrapper/test-helper/comments.data");
jest.mock("../actions/selected-comments-actions", () => {
  return {
    removeSelectedCommentAction: jest.fn(),
  };
});
jest.mock("react-redux", () => {
  return {
    useSelector: jest.fn(),
    useDispatch: jest.fn(),
  };
});
jest.mock("react-router-dom", () => {
  return {
    useHistory: jest.fn(),
  };
});

describe("SelectedCommentsPage", () => {
  const historyMock = jest.fn();
  const dispatchMock = jest.fn();
  beforeEach(() => {
    useHistory.mockReturnValue(historyMock);
    useDispatch.mockReturnValue(dispatchMock);
    useSelector.mockImplementation((selectorFn) =>
      selectorFn({ data: payload })
    );

    removeSelectedCommentAction.mockReturnValue(() => jest.fn());

    jest.clearAllMocks();
  });
  test("should match snapshot HomePage component", () => {
    const tree = renderer.create(<SelectedCommentsPage />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  test("should find table component when selectedComments > 0", () => {
    const tree = mount(<SelectedCommentsPage />);

    const table = tree.find("main TableComponent");
    console.log(table.prop("rows"));
    expect(table.prop("rows")).toEqual(
      payload.comments.filter((c) => c.selected)
    );
  });
  test("should not find table component when selectedComments === 0", () => {
    useSelector.mockImplementation((selectorFn) =>
      selectorFn({ data: { comments: [] } })
    );
    const tree = mount(<SelectedCommentsPage />);

    const table = tree.find("main TableComponent");
    expect(table).toEqual({});
  });
});
