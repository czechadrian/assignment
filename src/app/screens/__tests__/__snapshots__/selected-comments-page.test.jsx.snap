// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectedCommentsPage should match snapshot HomePage component 1`] = `
.c0 {
  width: 80%;
  margin: 0 10% 5%;
  border-bottom: 1px solid lightgray;
}

.c1:last-child {
  margin-bottom: 2.5%;
}

.c4 {
  align: right;
}

.c3 {
  aria-label: "simple table";
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 0 10% 0;
}

<main>
  <header
    className="c0"
  >
    <div
      className="MuiBottomNavigation-root c1"
    >
      <button
        className="MuiButtonBase-root MuiBottomNavigationAction-root Mui-selected"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiBottomNavigationAction-wrapper"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
            />
          </svg>
          <span
            className="MuiBottomNavigationAction-label Mui-selected"
          >
            Homepage
          </span>
        </span>
      </button>
    </div>
  </header>
  <main
    className="c2"
  >
    <div
      className="MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded"
    >
      <table
        className="MuiTable-root c3"
        role={null}
      >
        <thead
          className="MuiTableHead-root"
          role={null}
        >
          <tr
            className="MuiTableRow-root MuiTableRow-head"
            role={null}
          >
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head c4"
              scope="col"
            >
              Name
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head c4"
              scope="col"
            >
              E-mail
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head c4"
              scope="col"
            >
              First 20 characters
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head c4"
              scope="col"
            >
              Add/Remove to Selected
            </th>
          </tr>
        </thead>
        <tbody
          className="MuiTableBody-root"
          role={null}
        >
          <tr
            className="MuiTableRow-root"
            role={null}
          >
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body"
              role="cell"
              scope="row"
            >
              id labore ex et quam laborum
            </th>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body c4"
            >
              Eliseo@gardner.biz
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body c4"
            >
              laudantium enim quas
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body c4"
            >
              <button
                className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={0}
                type="button"
              >
                <span
                  className="MuiButton-label"
                >
                  Remove
                </span>
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </main>
</main>
`;
